components/p3a/brave_p3a_utils.h

```diff
@@ -0,0 +1,50 @@
+/* Copyright 2019 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_P3A_BRAVE_P3A_UTILS_H_
+#define BRAVE_COMPONENTS_P3A_BRAVE_P3A_UTILS_H_
+
+class PrefRegistrySimple;
+class PrefService;
+
+namespace brave_shields {
+
+// Note: these are APPEND-ONLY enumerations! Never remove any existing values,
+// as these enums are used to bucket a UMA histogram, and removing values
+// breaks that.
+enum ShieldsIconUsage {
+  kNeverClicked,
+  kClicked,
+  kShutOffShields,
+  kChangedPerSiteShields,
+  kSize,
+};
+
+// With the "Maybe" methods, we save latest value to local state and compare
+// new values with it. The idea is to write to a histogram only the highest
+// value (e.g. we are not interested in |kClicked| event if the user already
+// turned off shields. Since P3A sends only latest written values, these are
+// enough for our current goals.
+void MaybeRecordShieldsUsageP3A(ShieldsIconUsage usage,
+                                PrefService* local_state);
+}  // namespace brave_shields
+
+namespace brave {
+
+enum NTPCustomizeUsage {
+  kNeverOpened,
+  kOpened,
+  kOpenedAndEdited,
+  kCustomizeUsageMax
+};
+
+void MaybeRecordNTPCustomizeUsageP3A(NTPCustomizeUsage usage,
```

Comment:
What do you think of replacing this with a general utility like the following:
```
template<class Enumeration>
void RecordValueIfGreater(Enumeration value,
                          const char* hist_name,
                          const char* pref_name,
                          PrefService* local_state) {
  // May be null in tests.
  if (!local_state) {
    return;
  }
  const int usage_int = static_cast<int>(usage);
  int last_value = local_state->GetInteger(pref_name);
  if (last_value < usage_int) {
    base::UmaHistogramExactLinear(hist_name, usage_int,
                                  static_cast<int>(Enumeration::kSize));
    local_state->SetInteger(pref_name, usage_int);
  }
}
```
and moving `NTPCustomizeUsage` to corresponding components? I would really like to keep `components/p3a` unaware of other stuff like shields, NTP, etc.

Reply:
FWIW, I've tried to write a template function to utilise `UMA_HISTOGRAM` macro, but failed due to discrepancy in `histogram_macros.h`. Will try to upstream a fix :D

Reply:
This is definitely a nasty one - I would like to make more generic too- this is similar to what I had put together:
```diff

diff --git a/components/p3a/brave_p3a_utils.cc b/components/p3a/brave_p3a_utils.cc
index 033a1b5b806..d72a690cb87 100644
--- a/components/p3a/brave_p3a_utils.cc
+++ b/components/p3a/brave_p3a_utils.cc
@@ -44,7 +44,61 @@ void MaybeRecordNTPCustomizeUsageP3A(NTPCustomizeUsage usage,
   }
 }
 
+std::vector<std::string> prefs_to_register_;
+
+void RegisterP3ARecordedUsageInt(PrefRegistrySimple* local_state) {
+  // TODO: ..
+  for (auto pref : prefs_to_register_) {
+    LOG(ERROR) << "BSC]] demo] " << pref;
+    //local_state->RegisterIntegerPref(kShieldUsageStatus, -1);
+  }
+}
+
+template <typename T> class P3ARecordedUsageInt {
+ public:
+  P3ARecordedUsageInt(T max_value, std::string pref_name,
+      std::string event_name)
+      : enum_max_value_(max_value),
+        pref_name_(pref_name),
+        event_name_(event_name) {
+    DCHECK(std::is_enum<T>::value);
+    prefs_to_register_.push_back(pref_name_);
+  }
+
+  void MaybeRecordUsageIntP3A(T new_value, PrefService* local_state) {
+    // May be null in tests.
+    if (!local_state) {
+      return;
+    }
+    int last_value = local_state->GetInteger(pref_name_.c_str());
+    int new_value_int = static_cast<int>(new_value);
+    if (last_value < new_value_int) {
+      // UMA_HISTOGRAM_ENUMERATION(event_name.c_str(), new_value,
+      //     params.enum_max_value);
+      LOG(ERROR) << "BSC]]" << (int)enum_max_value_;
+      local_state->SetInteger(pref_name_.c_str(), new_value_int);
+    }
+  }
+
+ private:
+  T enum_max_value_;
+  std::string pref_name_;
+  std::string event_name_;
+};
+
 void RegisterP3AUtilsPrefs(PrefRegistrySimple* local_state) {
+  // example usage (this is what would go into an implementation)
+  P3ARecordedUsageInt<NTPCustomizeUsage> usage(
+      kCustomizeUsageMax,
+      kNTPCustomizeUsageStatus,
+      "Brave.NTP.CustomizeUsageStatus");
+  usage.MaybeRecordUsageIntP3A(kNeverOpened, NULL);
+
+  //...
+  RegisterP3ARecordedUsageInt(local_state);
+
+
+  //old code
   local_state->RegisterIntegerPref(kShieldUsageStatus, -1);
   local_state->RegisterIntegerPref(kNTPCustomizeUsageStatus, -1);
 }
```

I'll take another look at this soon

Reply:
Will try what you proposed @iefremov 😄 Much simpler